### PR TITLE
Zip

### PR DESCRIPTION
**Pull request by AELSchauer**
Adds ZipGenerator class (with placeholder test file, but no tests; how the heck do you test this???)
Updates some of the controller tests
Adds some new methods to the Upload and Folder models
Adds a fixtures folder for attachments to use for seeding and testing

**So, all the tests were passing, but when I merged master into this branch to resolve conflicts, a ton of tests started failing. I think it's better to wait until all our PRs are merged, and then make another branch to fix the tests.**

### CODE REVIEW

File: app/services/zip_generator.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      uploads.each do |upload|
+        if download_type == Folder
+          zipfile.add(upload.folio_filepath, upload.local_filepath)
+        elsif download_type == Upload
+          zipfile.add(upload.name, upload.local_filepath)
+        end
+      end
+    end
+  end
+
```
Comment (cmacaulay) [inline app/services/zip_generator.rb]: 👏 @AELSchauer can you walk us through some of this on Monday? Thank you so much for all of your hard work on this complicated logic.

File: spec/models/upload_spec.rb [hunk 2]
```diff
@@ -1,5 +1,87 @@
...
+        "attachment",
+        "user_#{file.user.id}",
+        "folder_#{file.folder.id}",
+        "upload_#{file.id}",
+        file.name
+      ).to_s
+
+      expect(file.local_filepath).to eq(actual_local_filepath)
+    end
+  end
```
Comment (cmacaulay) [inline spec/models/upload_spec.rb]: 👍 great test!

File: spec/models/folder_spec.rb [hunk 3]
```diff
@@ -20,4 +20,57 @@
...
+      end
+
+      it "in itself and a subfolder" do
+        user  = create(:user)
+        root  = create(:folder, user: user)
+        one   = create(:folder, user: user, parent: root)
+        two   = create(:folder, user: user, parent: one)
+        file1 = create(:upload, folder: root)
+        file2 = create(:upload, folder: one)
+        file3 = create(:upload, folder: two)
```
Comment (cmacaulay) [inline spec/models/folder_spec.rb]: it's really helpful seeing this factory pattern, sometimes it's hard to figure out how to generate them based off of their relationships - great work!

File: spec/services/zip_generator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+require "rails_helper"
+
+RSpec.describe ZipGenerator do
+  xit "need ZipGenerator tests" do
+  end
+end
```
Comment (AELSchauer) [inline spec/services/zip_generator_spec.rb]: Done.

File: app/services/zip_generator.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  def data
+    File.read(temp_file.path)
+  end
+
+  def close
+    temp_file.close
+    temp_file.unlink
+  end
+end
```
Comment (AELSchauer) [inline app/services/zip_generator.rb]: Done.

File: app/services/zip_generator.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  attr_accessor :temp_file
+
+  def initialize(selection)
+    @download_type = selection.class
+    @file_name     = "#{selection.name}.zip"
+    @uploads       = selection.all_uploads
+    @temp_file     = Tempfile.new(file_name)
+  end
+
+  def download
```
Comment (AELSchauer) [inline app/services/zip_generator.rb]: Done?

File: spec/controllers/folders/download_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+  describe "GET #index" do
+    it "returns http success" do
+      root = create(:folder)
+      get :index, params: { id: root.id }
+
+      expect(response).to have_http_status(:success)
+    end
+  end
+
```
Comment (AELSchauer) [inline spec/controllers/folders/download_controller_spec.rb]: Done.

File: spec/controllers/uploads/download_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+  describe "GET #index" do
+    it "returns http success" do
+      upload = create(:upload)
+      get :index, params: {id: upload.id}
+
+      expect(response).to have_http_status(:success)
+    end
+  end
+
```
Comment (AELSchauer) [inline spec/controllers/uploads/download_controller_spec.rb]: Done.

File: spec/controllers/uploads/download_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
+require 'rails_helper'
+
+RSpec.describe Uploads::DownloadController, type: :controller do
+
+  describe "GET #index" do
+    it "returns http success" do
+      upload = create(:upload)
+      get :index, params: {id: upload.id}
```
Comment (AELSchauer) [inline spec/controllers/uploads/download_controller_spec.rb]: Done.

File: spec/factories/uploads.rb [hunk 10]
```diff
@@ -1,8 +1,9 @@
 FactoryGirl.define do
   factory :upload do
-    sequence(:name) { |n| "File #{n}" }
-    content_type "application/pdf"
-    size 200
+    attachment { Rack::Test::UploadedFile.new(File.join(Rails.root, "spec", "support", "fixtures", "attachments", "femalecodercat.jpg"), "image/jpg") }
```
Comment (AELSchauer) [inline spec/factories/uploads.rb]: Added factories to ignore.

File: config/routes.rb [hunk 11]
```diff
@@ -5,23 +5,28 @@
...
-  # resources :folders, path: :f, only: [:show, :new, :create]
-  resources :uploads, path: :u, only: [:new, :create]
+  get "/f/:id/download", to: "folders/download#index", as: "folder_download"
+
+  # uploads, comments & download
+  resources :uploads, path: :u, only: [:show, :create]
+  get "/u/:id/download", to: "uploads/download#index", as: "upload_download"
   resources :uploads, path: :u, only: [:show] do
     resources :comments, only: [:create]
   end
```
Comment (cmacaulay) [inline config/routes.rb]: @AELSchauer I think you're very right in identifying this as an area that needs refactoring.

File: app/services/zip_generator.rb (hunk 5 above)
Comment (AELSchauer) [inline app/services/zip_generator.rb]: Done

File: app/services/zip_generator.rb (hunk 6 above)
Comment (AELSchauer) [inline app/services/zip_generator.rb]: Done.

File: spec/controllers/folders/download_controller_spec.rb (hunk 7 above)
Comment (AELSchauer) [inline spec/controllers/folders/download_controller_spec.rb]: Done.

File: spec/controllers/uploads/download_controller_spec.rb (hunk 8 above)
Comment (AELSchauer) [inline spec/controllers/uploads/download_controller_spec.rb]: Done.

File: spec/controllers/uploads/download_controller_spec.rb (hunk 9 above)
Comment (AELSchauer) [inline spec/controllers/uploads/download_controller_spec.rb]: Done.

File: app/controllers/folders/download_controller.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
+class Folders::DownloadController < ApplicationController
```
Comment (AELSchauer) [inline app/controllers/folders/download_controller.rb]: Changed enforced style in rubocop

File: app/controllers/uploads/download_controller.rb [hunk 13]
```diff
@@ -0,0 +1,10 @@
+class Uploads::DownloadController < ApplicationController
```
Comment (AELSchauer) [inline app/controllers/uploads/download_controller.rb]: Changed enforced style in rubocop

File: app/helpers/folders/download_helper.rb [hunk 14]
```diff
@@ -0,0 +1,2 @@
+module Folders::DownloadHelper
```
Comment (AELSchauer) [inline app/helpers/folders/download_helper.rb]: Changed enforced style in rubocop

File: app/helpers/uploads/download_helper.rb [hunk 15]
```diff
@@ -0,0 +1,2 @@
+module Uploads::DownloadHelper
```
Comment (AELSchauer) [inline app/helpers/uploads/download_helper.rb]: Changed enforced style in rubocop

File: app/models/upload.rb [hunk 16]
```diff
@@ -14,4 +14,23 @@ class Upload < ApplicationRecord
   validates :size, presence: true
   validates :folder_id, presence: true
```
Comment (cmacaulay) [inline app/models/upload.rb]: @AELSchauer would you agree these could be refactored into a decorator class?

File: app/services/zip_generator.rb [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    Zip::File.open(temp_file.path, Zip::File::CREATE) do |zipfile|
+      uploads.each do |upload|
+        if download_type == Folder
+          upload_path = upload.folio_filepath
+        elsif download_type == Upload
+          upload_path = upload.name
+        end
+        zipfile.add(upload_path, upload.local_filepath)
```
Comment (cmacaulay) [inline app/services/zip_generator.rb]: 👍

File: app/views/uploads/show.html.erb [hunk 18]
```diff
@@ -6,7 +6,12 @@
 
 <p><%= @upload.content_type %></p>
 
+<<<<<<< HEAD
```
Comment (cmacaulay) [inline app/views/uploads/show.html.erb]: @AELSchauer looks like there's a merge conflict here.

File: spec/factories/uploads.rb (hunk 10 above)
Comment (cmacaulay) [inline spec/factories/uploads.rb]: 👍

File: app/models/upload.rb (hunk 16 above)
Comment (AELSchauer) [inline app/models/upload.rb]: Can you refactor validators? I thought they would have to be on the model.

File: spec/models/folder_spec.rb (hunk 3 above)
Comment (AELSchauer) [inline spec/models/folder_spec.rb]: Thanks 👍

**Comment by AELSchauer** [comment]
@cmacaulay Yeah, ADD brain got distracted from fixing the merge conflicts onto hound violations. oops. 😬  

I did a search and I think I fixed them all in a more recent commit, but if you find any more, definitely let me know and I can fix them! 😄

File: config/routes.rb (hunk 11 above)
Comment (AELSchauer) [inline config/routes.rb]: Yeah, I have an idea to maybe do it the way reddit does it with their subreddits. Our routes could even be more standard, I hope!

**Pull request merged**